File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System;
using System.Collections.Generic;

class Solution {
    private bool[,] visited;
    private int[] dx = {1, -1, 0, 0};
    private int[] dy = {0, 0, 1, -1};
    public int solution(int[,] maps) {
        int n = maps.GetLength(0);
        int m = maps.GetLength(1);
        int answer = BFS(0, 0, n, m, maps);
        
        return answer;
    }
    private int BFS(int x, int y, int n, int m, int[,] maps)
    {
        if(maps[x, y] == 0 || maps[n - 1, m - 1] == 0) return -1;
        
        int[,] dist = new int[n, m];
        Queue<(int, int)> queue = new Queue<(int, int)>();
        
        dist[0, 0] = 1;
        queue.Enqueue((x, y));
        
        while(queue.Count > 0)
        {
            var (cx, cy) = queue.Dequeue();
            for(int i = 0; i < 4; i++)
            {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                if(nx >= 0 && nx < n && ny >= 0 && ny < m && maps[nx, ny] == 1 && dist[nx, ny] == 0)
                {
                    dist[nx, ny] = dist[cx, cy] + 1;
                    queue.Enqueue((nx, ny));
                }
            }
        }
        return dist[n - 1, m - 1] == 0 ? - 1 : dist[n - 1, m - 1];
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<bits/stdc++.h>
using namespace std;

int dy[] = {-1, 0, 1, 0};
int dx[] = {0, 1, 0, -1};
int y, x;
int solution(vector<vector<int>> maps)
{
    int n = maps.size();
    int m = maps[0].size();
    int visited[n][m];
    
    memset(visited, 0, sizeof(visited));
    
    queue<pair<int, int>> q;
    q.push({0, 0});
    visited[0][0] = 1;
    
    while(q.size())
    {
        tie(y, x) = q.front();
        q.pop();
        for(int i = 0; i < 4; i++)
        {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if(ny < 0 || nx < 0 || ny >= n || nx >= m || maps[ny][nx] == 0) continue;
            if(visited[ny][nx]) continue;
            
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny, nx});
        }
    }
    return visited[n - 1][m - 1] == 0 ? -1 : visited[n - 1][m - 1];
}